src/Compilers/Core/Portable/DiagnosticAnalyzer/ShadowCopyAnalyzerAssemblyLoader.cs

```diff
@@ -11,60 +11,69 @@
 using Roslyn.Utilities;
 using System.Collections.Immutable;
 using System.Reflection;
-
-#if NET
-using System.Runtime.Loader;
-#endif
+using System.Globalization;
 
 namespace Microsoft.CodeAnalysis
 {
-    internal sealed class ShadowCopyAnalyzerAssemblyLoader : AnalyzerAssemblyLoader
+    internal sealed class ShadowCopyAnalyzerPathResolver : IAnalyzerPathResolver
     {
         /// <summary>
         /// The base directory for shadow copies. Each instance of
-        /// <see cref="ShadowCopyAnalyzerAssemblyLoader"/> gets its own
+        /// <see cref="ShadowCopyAnalyzerPathResolver"/> gets its own
         /// subdirectory under this directory. This is also the starting point
         /// for scavenge operations.
         /// </summary>
         private readonly string _baseDirectory;
 
+        /// <summary>
+        /// The name of the shadow directory. This will be the same name as used for the <see cref="_mutex"/>
+        /// field. That allows other instances of this type to coordinate on active and inactive
+        /// shadow copy directories.
+        /// </summary>
+        private readonly string _shadowDirectoryName;
+
+        private readonly string _shadowDirectory;
+
+        /// <summary>
+        /// As long as this mutex is alive, other instances of this type will not try to clean
+        /// up the shadow directory.
+        /// </summary>
+        private readonly Mutex _mutex;
+
         internal readonly Task DeleteLeftoverDirectoriesTask;
 
+        private int _directoryCount;
+
+        private readonly ConcurrentDictionary<string, int> _directoryMap = new();
+
         /// <summary>
-        /// The directory where this instance of <see cref="ShadowCopyAnalyzerAssemblyLoader"/>
-        /// will shadow-copy assemblies, and the mutex created to mark that the owner of it is still active.
+        /// Map from the shadow assembly path to the original assembly path
         /// </summary>
-        private readonly Lazy<(string directory, Mutex)> _shadowCopyDirectoryAndMutex;
+        private readonly ConcurrentDictionary<string, string> _shadowToOriginalPathMap = new();
 
-        private readonly ConcurrentDictionary<Guid, Task<string>> _mvidPathMap = new ConcurrentDictionary<Guid, Task<string>>();
-        private readonly ConcurrentDictionary<(Guid, string), Task<string>> _mvidSatelliteAssemblyPathMap = new ConcurrentDictionary<(Guid, string), Task<string>>();
+        private readonly ConcurrentDictionary<string, Task> _copyMap = new();
 
         internal string BaseDirectory => _baseDirectory;
 
-        internal int CopyCount => _mvidPathMap.Count;
+        internal string ShadowDirectory => _shadowDirectory;
 
-#if NET
-        public ShadowCopyAnalyzerAssemblyLoader(string baseDirectory, ImmutableArray<IAnalyzerAssemblyResolver>? externalResolvers = null)
-            : this(null, baseDirectory, externalResolvers)
-        {
-        }
+        internal int CopyCount => _copyMap.Count;
```

Comment:
1. doc.
2. why is this needed?  seems odd to expose a count.  but maybe more of the pr will make this clearer.

Reply:
It's used in testing to validate that the shadow copies are occurring as planned. There is extensive tracknig of what assemblies where loaded, where did they load from and how often is shadow copying occurring. This is part of that validation.